homeassistant/components/iaqualink/light.py

```diff
@@ -0,0 +1,111 @@
+"""Support for Aqualink pool lights."""
+import logging
+
+from iaqualink import AqualinkLight, AqualinkLightEffect
+
+from homeassistant.components.light import (
+    ATTR_BRIGHTNESS,
+    ATTR_EFFECT,
+    DOMAIN,
+    SUPPORT_BRIGHTNESS,
+    SUPPORT_EFFECT,
+    Light,
+)
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.helpers.typing import HomeAssistantType
+
+from .const import DOMAIN as AQUALINK_DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+PARALLEL_UPDATES = 0
+
+
+async def async_setup_entry(
+    hass: HomeAssistantType, config_entry: ConfigEntry, async_add_entities
+) -> None:
+    """Set up discovered switches."""
+    devs = []
+    for dev in hass.data[AQUALINK_DOMAIN][DOMAIN]:
+        devs.append(HassAqualinkLight(dev))
+    async_add_entities(devs, True)
+
+
+class HassAqualinkLight(Light):
+    """Representation of a light."""
+
+    def __init__(self, dev: AqualinkLight):
+        """Initialize the light."""
+        Light.__init__(self)
+        self.dev = dev
+
+    @property
+    def name(self) -> str:
+        """Return the name of the light."""
+        return self.dev.label
+
+    @property
+    def is_on(self) -> bool:
+        """Return whether the light is on or off."""
+        return self.dev.is_on
+
+    async def async_turn_on(self, **kwargs) -> None:
+        """Turn on the light.
+
+        This handles brightness and light effects for lights that do support
+        them.
+        """
+        brightness = kwargs.get(ATTR_BRIGHTNESS, None)
+        effect = kwargs.get(ATTR_EFFECT, None)
+
+        # For now I'm assuming lights support either effects or brightness.
+        if effect:
+            effect = AqualinkLightEffect[effect].value
+            await self.dev.set_effect(effect)
+        elif brightness:
+            # Aqualink supports percentages in 25% increments.
+            pct = int(round(brightness * 4.0 / 255)) * 25
+            await self.dev.set_brightness(pct)
+        else:
+            await self.dev.turn_on()
+
+    async def async_turn_off(self, **kwargs) -> None:
+        """Turn off the light."""
+        await self.dev.turn_off()
+
+    @property
+    def brightness(self) -> int:
+        """Return current brightness of the light.
+
+        The scale needs converting between 0-100 and 0-255.
+        """
+        return self.dev.brightness * 255 / 100
+
+    @property
+    def effect(self) -> str:
+        """Return the current light effect if supported."""
+        return AqualinkLightEffect(self.dev.effect).name
+
+    @property
+    def effect_list(self) -> list:
+        """Return supported light effects."""
+        return list(AqualinkLightEffect.__members__.keys())
+
+    async def async_update(self) -> None:
+        """Update the internal state of the light.
+
+        This is currently a no-op since all devices get refreshed during the
+        main thermostat update.
```

Comment:
It's not good from an architecture standpoint to let one platform drive the update that other platforms rely on. We should probably move the system update to the central component.

How does the feedback from the system work when changing the state of a light? Is `self.dev.is_on` updated independently of `self.dev.system.update` in the climate entity?

Reply:
Update() on the climate entiry will trigger the API requests which will in turn update the state of all the iaqualink devices. Is_on() is just a wrapper function to return the internal state of each device. It doesn't in itself query anything. Does that make more sense?

Reply:
So there's no feedback from the system after turning a light on, ie we don't know if the light was actually turned on?

Reply:
The feedback is the response to the API call. It's parsed and updates the state of the devices (all of them, not just the light that's been turned on) inside the library. There's no subscription or notification so we still rely on polling in case another application makes changes to the system.

Reply:
Ok, so `self.dev.is_on` will return the correct updated state after turning a light on?

Reply:
Correct.

Reply:
Good!

I suggest we just move the system update call from the climate platform to the component. Schedule the update with `event.async_track_time_interval`. The callback needs to accept a time parameter.

Reply:
Ask and you shall receive! PR at home-assistant/home-assistant#26505.

Reply:
That dependent diff landed, code has been rebased and updated!